File: src/main/java/org/moinex/repository/InvestmentPerformanceSnapshotRepository.java
/*
 * Filename: InvestmentPerformanceSnapshotRepository.java
 * Created on: February 17, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.repository;

import java.util.List;
import java.util.Optional;
import org.moinex.model.InvestmentPerformanceSnapshot;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

@Repository
public interface InvestmentPerformanceSnapshotRepository
        extends JpaRepository<InvestmentPerformanceSnapshot, Integer> {

    Optional<InvestmentPerformanceSnapshot> findByMonthAndYear(Integer month, Integer year);

    @Query(
            "SELECT ips FROM InvestmentPerformanceSnapshot ips ORDER BY ips.year ASC, ips.month"
                    + " ASC")
    List<InvestmentPerformanceSnapshot> findAllOrderedByDate();

    @Modifying
    @Transactional
    @Query("DELETE FROM InvestmentPerformanceSnapshot")
    void deleteAll();

    boolean existsByMonthAndYear(Integer month, Integer year);
}

File: src/main/kotlin/org/moinex/repository/InvestmentPerformanceSnapshotRepository.kt
/*
 * Filename: InvestmentPerformanceSnapshotRepository.kt (original filename: InvestmentPerformanceSnapshotRepository.java)
 * Created on: February 17, 2026
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 07/03/2026
 */

package org.moinex.repository

import org.moinex.model.InvestmentPerformanceSnapshot
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Modifying
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional
import java.util.Optional

@Repository
interface InvestmentPerformanceSnapshotRepository : JpaRepository<InvestmentPerformanceSnapshot, Int> {
    fun findByMonthAndYear(
        month: Int,
        year: Int,
    ): Optional<InvestmentPerformanceSnapshot>

    @Query(
        "SELECT ips " +
            "FROM InvestmentPerformanceSnapshot ips " +
            "ORDER BY ips.year ASC, ips.month ASC",
    )
    fun findAllOrderedByDate(): List<InvestmentPerformanceSnapshot>

    @Modifying
    @Transactional
    @Query("DELETE FROM InvestmentPerformanceSnapshot")
    override fun deleteAll()

    fun existsByMonthAndYear(
        month: Int,
        year: Int,
    ): Boolean
}
